iree/compiler/Dialect/Flow/Transforms/HLOToHLOPreprocessing.cpp

```diff
@@ -324,6 +326,249 @@ class AdjustDepthwiseFilterShape : public OpRewritePattern<mhlo::ConvOp> {
   }
 };
 
+// Rewrites rank-3 mhlo.dot general so lhs contraction dimension is
+// inner most (2) and rhs contraction dimension is dim right after batch
+// dimension. The pattern inserts transposes so the dot_general always has the
+// form: {batch_dim, parallel, contraction}.{batch_dim, contraction, parallel}
+class TransposeRank3GenericDotGeneral
+    : public OpRewritePattern<mhlo::DotGeneralOp> {
+ public:
+  using OpRewritePattern<mhlo::DotGeneralOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(mhlo::DotGeneralOp op,
+                                PatternRewriter &rewriter) const override {
+    auto lhsShapeType = op.lhs().getType().dyn_cast_or_null<RankedTensorType>();
+    auto rhsShapeType = op.rhs().getType().dyn_cast_or_null<RankedTensorType>();
+    auto resultType =
+        op.getResult().getType().dyn_cast_or_null<RankedTensorType>();
```

Comment:
`dyn_cast_or_null` would accept nullptr and I think the type would never be a nullptr, so `dyn_cast` is okay.

Reply:
+1 `dyn_cast` is fine. Actually `dyn_cast_or_null` here is weird. Not sure how it works at all.

Reply:
dyn_cast_or_null works because it does something like `return (!this || !isa<T>) ? null : cast<T>(this) ` but as base type will never be null so `dyn_cast` is good here,  miss understood the method from the name too :)

Reply:
TIL you can handle `this` being empty within a method of that class. Still seems odd to me.

Reply:
Definitely not just `this` =) `return (!this->imp_ptr || !isa<T>) ? null : cast<T>(this->imp_ptr)` =).

Reply:
Ah that makes more sense! THanks for the info.